Display receipts in chat

Li357 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/Expensify/issues/295245

We're bringing SmartScan receipts to NewDot!

### Tests

Before testing and before the Auth/Web-E changes make it to staging, please pull my branches https://github.com/Expensify/Auth/pull/8494 and https://github.com/Expensify/App/pull/24235 to test! Remember to rebuild Auth + restart.

Note that for all tests: adding money requests to an IOU report that already exists via Global Create creates an extra IOU  report action item, because the reportID is not in the URL when going thru Global Create. This is not in scope here, but we can fix in a later PR. For everything except for Test 1 step 1, use the + button the chat to create requests.

#### Test 1

0. Log into user A with no previous money requests with B
1. Click on Global Create > Request Money
2. Choose scan and upload an image
3. Assign the request to another known user B (you've chatted with them before)
4. Verify that a new Report Preview is created, with subtitle "Receipt" and headline "Scan in progress…"
5. Verify that the Report Preview is a whisper with "Only visible to you"
6. Verify that the local receipt image is immediately visible, but reloads to the remote image when the API responds
7. Verify that clicking on the Report Preview navigates to the IOU Report
8. Verify that in the IOU Report, the receipt also shows as "Receipt" and "Scan in progress…"
9. Verify that clicking on the IOU action brings up the Money Request View with "Receipt scan in progress…" in the LHN and a status bar under the Header with "Scanning…" and "Only you can see this receipt when it's scanning. Check back later or enter the details now."
10. Verify that clicking on the receipt image in this view opens up a modal with title "Receipt"
11. Log into user B and verify that you cannot see the Report Preview in the Chat & IOU reports

#### Test 2

1. Log back into user A
2. Type in the Chat Report with the report preview
3. Add a manual request
4. Verify that the Report Preview comes to the front of the chat report
5. Verify that the Report Preview has subtitle "2 requests, 1 scanning"
6. Verify that the Report Preview is no longer a whisper
7. Add two more receipts and verify that all three appear in the Report Preview with subtitle "4 requests, 3 scanning"
8. Add another receipt and give it a description "Foobar" (for test 3), and verify that only the three most recent appear in the Report Preview, and a +1 indicator is shown on the last receipt image

### Test 3

0. Stay logged in as user A in NewDot
1. Wait for the receipt to finish scanning
2. Verify that the report preview is not longer a whisper
3. Verify that the report preview subtitle is now "4 requests, 3 scanning"
4. Verify that in the IOU report, the last receipt is no longer a whisper
5. Verify that in the Money Request View, "Receipt scan in progress" is no longer in the LHN nor is there a status bar underneath the header

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Most functionality should work offline, since we optimistically create/update the report previews and IOU report actions. In particular:

Test 1: steps 0-10 on a new account C
Test 2: all steps

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
13. Upload an image via copy paste
14. Verify a modal appears displaying a preview of that image
--->

Same as manual tests above

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: https://expensify.slack.com/archives/C21FRDWCV/p1692217673735579
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



https://github.com/Expensify/App/assets/12980971/32b863ff-981e-4419-acd9-b3a70a2961ca


https://github.com/Expensify/App/assets/12980971/26b92c72-5336-4f83-900c-428899689337


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->


https://github.com/Expensify/App/assets/12980971/226422c8-4b5d-4879-81d4-015c5b855212


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->



https://github.com/Expensify/App/assets/12980971/d973c35d-8263-49cf-a7ec-f5dee65c8cdc


</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

 

https://github.com/Expensify/App/assets/12980971/c0ae650c-b73b-4585-a56e-084b030a8166


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

Commit:
Disable copy to clipboard for report previews

Commit:
Show receipt image

Commit:
Fix receipt image loading

Commit:
Begin styling for multiple receipts

Commit:
Fit images to container

Commit:
Polish preview box on hover

Commit:
Only apply full width when fitting container

Commit:
Add scanning in progress

Commit:
Fix thumbnail and add correct optimistic filename

Commit:
Fix filename when not present

Commit:
Rename IOUPreview to MoneyRequestPreview

Commit:
Implement whispers

Commit:
Update MoneyRequestPreview for receipts

Commit:
Refactor thumbnail/image for receipts

Commit:
Update request header, disable receipt modal, update image URI blob check

Commit:
Update scan in progress string

Commit:
Add receipt in money request view

Commit:
Add receipt whisper to details

Commit:
Move receipt whisper to header

Commit:
Show merchant and number of requests in report preview

Commit:
Add receipt preview modal to money request view

Commit:
Add receipt title and options for preview modal

Commit:
Add additional money requests indicator and refactor

Commit:
Add number of requests scanning in report preview subtitle

Commit:
Refactor image component and reuse

Commit:
Merge main

Commit:
Fix refactor of MoneyRequestImage

Commit:
Show comment and request count info correctly

Commit:
Fix linting and dependency cycle

Commit:
Run prettier

Li357 commented:
@Santhosh-Sellavel We're building out a new feature that involves a few backend changes. I'll reassign once we're ready to have external eyes on it!

Commit:
Update optimistic report preview data

Commit:
Add optimistic data for updating the report preview

Commit:
Remove MoneyRequestImage and update optimistic data

Commit:
Show number of requests before comment, update optimistic data

MariaHCD commented:
Following `Test 2`, looks good!

https://github.com/Expensify/App/assets/12268372/a29c6893-b600-4ac7-b6e6-b9c8ee0fda85

Similar to my previous test above, the reports show up for the payer:

<img width="1215" alt="Screenshot 2023-08-15 at 4 46 15 PM" src="https://github.com/Expensify/App/assets/12268372/455af53e-baec-4364-8bc3-ce341cdf9e4d">

Li357 commented:
@MariaHCD Thanks a ton for testing! The whispers are working as expected for me (and note that whispers will be a little off until we update the visible reply and request account to be scoped to user which @luacmartins will do). Did you pull latest Web-E and Auth and rebuild/restart?

The receipt drag functionality is frustrating. I can't seem to debug this, I'll ask @AndrewGable about it.

luacmartins commented on src/components/ReportActionItem/MoneyRequestView.js:
```diff
@@ -22,7 +22,12 @@ import DateUtils from '../../libs/DateUtils';
 import useLocalize from '../../hooks/useLocalize';
+import * as TransactionUtils from '../../libs/TransactionUtils';
+import * as ReceiptUtils from '../../libs/ReceiptUtils';
+import withLocalize from '../withLocalize';
```
We're already using `useLocalize` in this component. Let's stick to that.

luacmartins commented on src/components/ReportActionItem/ReportActionItemImages.js:
```diff
@@ -0,0 +1,82 @@
+    ).isRequired,
+
+    /** optional: max number of images to show in the row if different than images length */
+    // eslint-disable-next-line react/require-default-props
```
We should add defaults for these instead of disabling the lint rule.

Li357 replied:
The goal is to have size and total get their "default" value from the image prop length, but you can't set default props from other props AFAIK. This was the best I could come up with, and I can add a comment to note this. [Context](https://expensify.slack.com/archives/C05DWUDHVK7/p1692046582429169?thread_ts=1691789223.761189&cid=C05DWUDHVK7)

s77rt replied:
@Li357 Tagging for visibility

luacmartins commented on src/components/ReportActionItem/ReportActionItemImages.js:
```diff
@@ -0,0 +1,82 @@
+        }),
+    ).isRequired,
+
+    /** optional: max number of images to show in the row if different than images length */
```
We can remove optional from the comments. The lack of `isRequired` already indicates that it's optional.

```suggestion
    /** optional: max number of images to show in the row if different than images length */
```

luacmartins commented on src/components/ReportActionItem/ReportActionItemImages.js:
```diff
@@ -0,0 +1,82 @@
+                        style={[styles.reportActionItemImage, props.hoverStyle]}
+                    >
+                        {thumbnail ? (
+                            <RenderHTML
+                                html={`
+                                    <img
+                                        src="${thumbnail}"
+                                        data-expensify-source="${image}"
+                                        data-expensify-fit-container="true"
+                                        data-expensify-preview-modal-disabled="true"
+                                    />
+                            `}
+                            />
```
I see that we are using this in at least two places. Maybe we should extract it into it's own component.

luacmartins commented on src/components/ReportActionItem/ReportPreview.js:
```diff
@@ -114,11 +133,17 @@ function ReportPreview(props) {
 
+    let previewMessage;
     const managerName = ReportUtils.isPolicyExpenseChat(props.chatReport) ? ReportUtils.getPolicyName(props.chatReport) : ReportUtils.getDisplayNameForParticipant(managerID, true);
+    if (isScanning) {
+        previewMessage = props.translate('common.receipt');
+    } else {
+        previewMessage = props.translate(iouSettled || props.iouReport.isWaitingOnBankAccount ? 'iou.payerPaid' : 'iou.payerOwes', {payer: managerName});
+    }
```
Can we create a function to return `previewMessage`? I think that'd be cleaner and more scalable

Commit:
Add file URI when using drag and drop

Li357 commented:
Fixed the drag and drop issue! Seems like a file URI was not generated

luacmartins commented on src/components/AttachmentModal.js:
```diff
@@ -318,12 +322,30 @@ function AttachmentModal(props) {
                     shouldShowBackButton={props.isSmallScreenWidth}
+                    shouldShowThreeDotsButton={isAttachmentReceipt}
+                    threeDotsMenuItems={[
+                        {
+                            icon: Expensicons.Camera,
+                            text: props.translate('common.replace'),
+                            onSelected: () => {}, // TODO
+                        },
```
NAB we have a specific [issue](https://github.com/Expensify/Expensify/issues/295248) for adding the replace functionality. Maybe we should leave adding this stuff for that issue.

luacmartins commented on src/libs/ReportUtils.js:
```diff
@@ -2014,10 +2022,15 @@ function buildOptimisticModifiedExpenseReportAction(transactionThread, oldTransa
+function updateReportPreview(iouReport, reportPreviewAction, comment = '', transaction = undefined) {
+    const hasReceipt = TransactionUtils.hasReceipt(transaction);
+    const lastReceiptTransactionIDs = lodashGet(reportPreviewAction, 'childLastReceiptTransactionIDs', '');
+    const previousIDs = lastReceiptTransactionIDs.split(',').slice(0, 2);
```
```suggestion
    const previousTransactionIDs = lastReceiptTransactionIDs.split(',').slice(0, 2);
```

luacmartins commented on src/libs/fileDownload/FileUtils.js:
```diff
@@ -152,7 +152,8 @@ const readFileAsync = (path, fileName) =>
             })
             .then((blob) => {
-                const file = new File([blob], cleanFileName(fileName));
+                const cleanName = cleanFileName(fileName);
+                const file = new File([blob], cleanName);
```
Why is this change necessary?

Li357 replied:
I think this was a change I had forgotten to revert

luacmartins commented on src/pages/iou/ReceiptSelector/index.js:
```diff
@@ -99,8 +99,7 @@ function ReceiptSelector(props) {
 
-        const filePath = URL.createObjectURL(file);
-        IOU.setMoneyRequestReceipt(filePath, file.name);
+        IOU.setMoneyRequestReceipt(file.uri, file.name);
```
Why is this change necessary?

luacmartins reviewed (changes requested):
Left a few comments

mountiny commented on src/libs/ReportActionsUtils.js:
```diff
@@ -580,6 +595,93 @@ function isMessageDeleted(reportAction) {
+ * @param {Object} iouReportAction
+ * @returns {Object}
+ */
+function getTransaction(iouReportAction) {
+    const transactionID = lodashGet(iouReportAction, ['originalMessage', 'IOUTransactionID']);
+    return allTransactions[transactionID] || {};
+}
+
```
We already have this method https://github.com/Expensify/App/blob/61ab4bcf100500ce49bf1e2dfaa128f9511ea941/src/libs/TransactionUtils.js#L94-L102 we should reuse it

Li357 replied:
Ah, this didn't get caught by the merge.

mountiny commented on src/libs/ReportActionsUtils.js:
```diff
@@ -37,6 +39,19 @@ Onyx.connect({
     },
 });
 
+const allTransactions = {};
+Onyx.connect({
+    key: ONYXKEYS.COLLECTION.TRANSACTION,
+    callback: (actions, key) => {
+        if (!key || !actions) {
+            return;
+        }
+
+        const transactionID = CollectionUtils.extractCollectionItemID(key);
+        allTransactions[transactionID] = actions;
+    },
+});
+
```
I think we should avoid adding this here, we are pulling all the transactions in the `TransactionUtils` here https://github.com/Expensify/App/blob/61ab4bcf100500ce49bf1e2dfaa128f9511ea941/src/libs/TransactionUtils.js#L10-L21 and `getReportPreviewTransactionsWithReceipts` should be moved there as it does not new allReportActions. This will keep it cleaner imho

Li357 commented on src/components/MoneyRequestHeader.js:
```diff
@@ -91,6 +95,9 @@ function MoneyRequestHeader(props) {
                     shouldShowBackButton={props.isSmallScreenWidth}
                     onBackButtonPress={() => Navigation.goBack(ROUTES.HOME, false, true)}
                     shouldShowBorderBottom
+                    shouldShowStatusBar={isScanning}
```
Agreed, I should've clarified with you on the scope of the "Display receipts in chat" part of the design doc.

Commit:
Address comments and update comments and docs

Li357 commented on src/libs/ReportActionsUtils.js:
```diff
@@ -580,6 +595,93 @@ function isMessageDeleted(reportAction) {
+ * @param {Object|null} reportAction
+ * @returns {Boolean}
+ */
+function hasReadyMoneyRequests(reportAction) {
```
Updated the docs on the function. Do you have any ideas for a better name with the new description?

mountiny replied:
I think its better to have longer but more descriptive names  but I am alo inding it hard to come up with something, its odd to look for `NO or at least one` 😂

MariaHCD replied:
From the conversation in the DD: https://docs.google.com/document/d/1Il9y1287B6tq0FB5ziXnhBcwn-Lw-n_rSmllZ-AyUVs/edit?disco=AAAA4PwHkcY

`areAllRequestsBeingSmartscanned` should be more clear!

mountiny replied:
I dont think we came to a good conclusion, maybe lets just leave it for later

Commit:
Separate status bar from header

Commit:
Refactor image to different component and fix linting

Gonals commented:
Hey @Li357, I talked with Tom about how receipts should look on web in the `edit` screen and here's the deal:
![Uploading image-2023-08-16 09_40_35.244.png…]()

MariaHCD commented:
@Gonals maybe we should consider adjusting the receipt thumbnail for web in a follow up PR? So we don't increase the scope here.

luacmartins commented:
@Li357 It seems like the button margin/padding is incorrect
<img width="496" alt="Screenshot 2023-08-16 at 8 45 06 AM" src="https://github.com/Expensify/App/assets/22219519/0b476102-93b8-43fd-993f-98309f56205b">

Commit:
Rename to areAllRequestsBeingSmartScanned

Commit:
Fix settlement button padding and typos

Li357 committed:
Merge branch 'main' into andrewli-displayreceipts

Commit:
Add maxWidth to receipt preview in money request view

Commit:
Add arrow right on report preview/iou preview

Commit:
Remove test code

Commit:
Attach URI to file for iOS uploads, change optimistic update for whisper

Li357 commented:
There does seem to be a whisper bug that we can't immediately fix. Whispers aren't removed correctly because in the backend we remove the property, and Onyx does not wipe out a property when we remove when merging. I'll fix this as part of [Add SmartScan in progress](https://github.com/Expensify/Expensify/issues/301367)

mountiny commented:
@Li357 Got bunch of merge conflicts now

Commit:
Merge main

Commit:
Fix linting and prettier

Commit:
Remove identifier

s77rt commented:
Bug: After deleting a receipt request the transaction is still there, this made the ReportPreview looks like it has an image but it's empty

Bug?: If we end up storing an invalid url in onyx the console will yield a ton of error messages and slow the app (This bug may only be reproducible before Web-E 38478)


https://github.com/Expensify/App/assets/16493223/b5a37aa9-d624-4fc7-874a-09d662b0dff1



Both bugs seems local only, after clearing onyx, sign out and sign in it seems fixed.

_NEED TO RETEST_

s77rt commented on src/libs/fileDownload/FileUtils.js:
```diff
@@ -154,6 +154,7 @@ const readFileAsync = (path, fileName) =>
             .then((blob) => {
                 const file = new File([blob], cleanFileName(fileName));
                 file.source = path;
+                file.uri = path;
```
It does not look that we are using the `uri` anywhere, any reason we are adding it?

Li357 replied:
I'll add a comment. On iOS, File objects do not seem to have a `uri` property, so uploading to the backend fails.

Li357 commented on src/components/MoneyRequestHeader.js:
```diff
@@ -90,6 +117,7 @@ function MoneyRequestHeader(props) {
                     personalDetails={props.personalDetails}
                     shouldShowBackButton={props.isSmallScreenWidth}
                     onBackButtonPress={() => Navigation.goBack(ROUTES.HOME, false, true)}
+                    statusBar={isScanning ? getStatusBar() : null}
                     shouldShowBorderBottom
```
Hmm. This is a new pattern we're building out, and I suspect more screens will want to supply a statusbar. Instead of inlining, maybe we can move this to a separate MoneyRequestHeaderStatusBar component?

s77rt replied:
MoneyRequestHeaderStatusBar looks good

Li357 commented on src/languages/en.js:
```diff
@@ -274,8 +275,8 @@ export default {
         markAsRead: 'Mark as read',
         editComment: 'Edit comment',
-        deleteAction: ({action}) => `Delete ${ReportActionsUtils.isMoneyRequestAction(action) ? 'request' : 'comment'}`,
-        deleteConfirmation: ({action}) => `Are you sure you want to delete this ${ReportActionsUtils.isMoneyRequestAction(action) ? 'request' : 'comment'}?`,
+        deleteAction: ({isMoneyRequest}) => `Delete ${isMoneyRequest ? 'request' : 'comment'}`,
+        deleteConfirmation: ({isMoneyRequest}) => `Are you sure you want to delete this ${isMoneyRequest ? 'request' : 'comment'}?`,
```
Good catch. The reason I made this change was to remove a dependency cycle, because requiring ReportActionsUtils in en.js/es.js causes a huge cycle. I'll see if I can change this back

Li357 replied:
Refactored so that ReportActionsUtils does not require ReceiptUtils!

Li357 commented on src/components/ReportActionItem/ReportActionItemImage.js:
```diff
@@ -0,0 +1,51 @@
+function ReportActionItemImage({thumbnail, image, enablePreviewModal}) {
+    if (thumbnail) {
+        return (
+            <RenderHTML
+                html={`
+                    <img
+                        src="${thumbnail}"
+                        data-expensify-source="${image}"
+                        data-expensify-fit-container="true"
+                        data-expensify-preview-modal-disabled="${!enablePreviewModal}"
+                    />
+            `}
+            />
```
I agree that ThumbnailImage would simplify ReportPreview and MoneyRequestPreview, but we would probably use RenderHTML for the MoneyRequestView.

s77rt replied:
Is this due to `enablePreviewModal`? It may be best to move the logic in `ImageRenderer` to another component and use it there and in here

(not a blocker - possibly a follow up)

Li357 replied:
I'd also agree, but probably a follow up where we can move the logic out and remove `data-expensify-fit-container`. For now, let's go with this.

Li357 commented on src/libs/ReceiptUtils.js:
```diff
@@ -23,6 +28,45 @@ const validateReceipt = (file) => {
+ */
+function getThumbnailAndImageURIs(path, filename) {
+    // For local files, we won't have a thumbnail yet
+    if (path.startsWith('blob:') || path.startsWith('file:')) {
```
This shouldn't happen with Web-E deploy 👍

s77rt commented on src/styles/styles.js:
```diff
@@ -3665,6 +3669,61 @@ const styles = {
+        width: '100%',
+    },
+
+    reportPreviewBoxHoverBorder: {
+        borderColor: themeColors.border,
+        backgroundColor: themeColors.border,
+    },
```
This style is same as `moneyRequestPreviewBoxHover`

s77rt commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -468,12 +468,21 @@ function ReportActionItem(props) {
-                        shouldShowHorizontalRule={!props.shouldHideThreadDividerLine}
-                    />
-                </OfflineWithFeedback>
+                <ShowContextMenuContext.Provider
+                    value={{
+                        anchor: popoverAnchorRef,
+                        report: props.report,
+                        action: props.action,
+                        checkIfContextMenuActive: toggleContextMenuFromActiveReportAction,
+                    }}
+                >
```
NAB. What case is this for?

Li357 replied:
We need the context provider to show the attachment preview modal correctly

Li357 commented on src/libs/ReportActionsUtils.js:
```diff
@@ -566,6 +568,69 @@ function isMessageDeleted(reportAction) {
+    // If a report preview has at least one manual request or at least one scanned receipt
+    if (isReportPreviewAction(reportAction)) {
+        const transactions = TransactionUtils.getReportPreviewTransactionsWithReceipts(reportAction);
+        // If we have more requests than requests with receipts, we have some manual requests
+        if (getNumberOfMoneyRequests(reportAction) > transactions.length) {
+            return true;
+        }
+        return _.some(transactions, (transaction) => !ReceiptUtils.isBeingScanned(transaction.receipt));
```
Good catch. We recently did a name change for the function and I forgot to update the comments/usages to reflect this.

Li357 commented on src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js:
```diff
@@ -48,7 +48,7 @@ const customHTMLElementModels = {
 };
 
-const defaultViewProps = {style: [styles.alignItemsStart, styles.userSelectText]};
+const defaultViewProps = {style: [styles.alignItemsStart, styles.userSelectText, styles.w100, styles.h100]};
```
I don't think this is easily doable without refactoring a lot of ReportActionItemImage's use of RenderHTML. I can do this in a followup PR.

dangrous replied:
This is the line that causes https://github.com/Expensify/App/issues/25678 unfortunately, can we think of a way to update it without refactoring everything else?

s77rt commented on src/libs/TransactionUtils.js:
```diff
@@ -157,6 +167,27 @@ function getCreated(transaction) {
+        transactionIDs,
+        (transactions, transactionID) => {
+            const transaction = getTransaction(transactionID);
+            if (!_.isEmpty(transaction)) {
```
```suggestion
            if (hasReceipt(transaction)) {
```

Commit:
Address comments

s77rt commented on src/components/MoneyRequestHeader.js:
```diff
@@ -117,8 +93,7 @@ function MoneyRequestHeader(props) {
                     onBackButtonPress={() => Navigation.goBack(ROUTES.HOME, false, true)}
-                    statusBar={isScanning ? getStatusBar() : null}
-                    shouldShowBorderBottom
+                    statusBar={isScanning ? <MoneyRequestHeaderStatusBar /> : null}
```
We still don't need a `statusBar` prop. It's better to not create a prop that may never get used. `MoneyRequestHeaderStatusBar` belongs to `MoneyRequestHeader` more than `HeaderWithBackButton`.

```js
<HeaderWithBackButton />
<MoneyRequestHeaderStatusBar />
```
Looks better

s77rt commented on src/components/MoneyRequestHeaderStatusBar.js:
```diff
@@ -0,0 +1,37 @@
+
+MoneyRequestHeaderStatusBar.displayName = 'MoneyRequestHeaderStatusBar';
+
+export default MoneyRequestHeaderStatusBar;
```
Missing newline (prettier?)

s77rt commented on src/libs/ReportActionsUtils.js:
```diff
@@ -566,6 +568,69 @@ function isMessageDeleted(reportAction) {
+        if (getNumberOfMoneyRequests(reportAction) > transactions.length) {
+            return false;
+        }
+        return _.all(transactions, (transaction) => !ReceiptUtils.isBeingScanned(transaction.receipt));
```
```suggestion
        return _.all(transactions, (transaction) => ReceiptUtils.isBeingScanned(transaction.receipt));
```

s77rt commented on src/libs/ReportActionsUtils.js:
```diff
@@ -566,6 +568,69 @@ function isMessageDeleted(reportAction) {
+    // If a money request action is not a scanning receipt
+    if (isMoneyRequestAction(reportAction)) {
+        const transaction = TransactionUtils.getTransaction(reportAction.originalMessage.IOUTransactionID);
+        return TransactionUtils.hasReceipt(transaction) && !ReceiptUtils.isBeingScanned(transaction.receipt);
```
```suggestion
        return TransactionUtils.hasReceipt(transaction) && ReceiptUtils.isBeingScanned(transaction.receipt);
```

s77rt commented on src/components/MoneyRequestHeader.js:
```diff
@@ -70,6 +71,8 @@ function MoneyRequestHeader(props) {
         setIsDeleteModalVisible(false);
     }, [parentReportAction, setIsDeleteModalVisible]);
 
+    const isScanning = !ReportActionsUtils.areAllRequestsBeingSmartScanned(parentReportAction);
```
```suggestion
    const isScanning = ReportActionsUtils.areAllRequestsBeingSmartScanned(parentReportAction);
```

Can you please double check the usage of `areAllRequestsBeingSmartScanned` in other places

Commit:
Move status bar component and fix conditions

Commit:
Fix prettier

s77rt commented:
Bug?: As user B I can see the receipt image in the preview but it's supposed to only be visible to user A

![Screenshot from 2023-08-17 18-41-29](https://github.com/Expensify/App/assets/16493223/9f9455cb-4dd0-42f9-af31-74e782825aea)

Commit:
Remove dependency cycle by moving getNumberOfScanningReceipts and areAllReceiptsBeingSmartScanned to ReportUtils

s77rt commented:
Bug: After adding more receipts, user B can see all the receipts and it shows a very confusing message "Only visible to user A" but you as user B can see the receipt as well

![Screenshot from 2023-08-17 18-47-57](https://github.com/Expensify/App/assets/16493223/00ff8d33-9f4d-49a2-9ee1-86fe2fe16fef)

s77rt commented:
Bug(backend?): The detected amount is negative
![Screenshot from 2023-08-17 18-53-29](https://github.com/Expensify/App/assets/16493223/b5bed23c-26d2-4f68-9f4f-81f3be40ff8e)
![9C5Q9Rt](https://github.com/Expensify/App/assets/16493223/a5e3934e-83c2-4b26-a4f8-af63c38d9359)

Commit:
Update Spanish translations

Commit:
Fix typo

Li357 commented on src/components/ReportActionItem/ReportPreview.js:
```diff
@@ -132,46 +161,55 @@ function ReportPreview(props) {
-                    </View>
-                    {showComment && (
-                        <View style={[styles.flexRow]}>
-                            <View style={[styles.flex1]}>
-                                <Text style={[styles.mt1, styles.colorMuted]}>
-                                    {moneyRequestCount > 1 ? props.translate('iou.requestCount', {count: moneyRequestCount}) : moneyRequestComment}
-                                </Text>
+                        <View style={styles.flexRow}>
+                            <View style={[styles.flex1, styles.flexRow, styles.alignItemsCenter]}>
+                                <Text style={styles.textHeadline}>{getDisplayAmount()}</Text>
+                                {ReportUtils.isSettled(props.iouReportID) && (
+                                    <View style={styles.defaultCheckmarkWrapper}>
+                                        <Icon
+                                            src={Expensicons.Checkmark}
+                                            fill={themeColors.iconSuccessFill}
+                                        />
+                                    </View>
+                                )}
                             </View>
                         </View>
-                    )}
-                    {shouldShowSettlementButton && (
-                        <SettlementButton
-                            currency={props.iouReport.currency}
-                            policyID={props.iouReport.policyID}
-                            chatReportID={props.chatReportID}
-                            iouReport={props.iouReport}
-                            onPress={(paymentType) => IOU.payMoneyRequest(paymentType, props.chatReport, props.iouReport)}
-                            enablePaymentsRoute={ROUTES.BANK_ACCOUNT_NEW}
-                            addBankAccountRoute={bankAccountRoute}
-                            style={[styles.requestPreviewBox]}
-                        />
-                    )}
+                        {hasReceipts && !isScanning && (
+                            <View style={styles.flexRow}>
+                                <View style={[styles.flex1, styles.flexRow, styles.alignItemsCenter]}>
+                                    <Text style={[styles.textLabelSupporting, styles.mb1, styles.lh16]}>{previewSubtitle || moneyRequestComment}</Text>
+                                </View>
+                            </View>
+                        )}
```
I'm going to resolve this for now. The subtitle shown in the card is going to change with Merchant and Comment changes coming later!

s77rt commented:
> Add another receipt and give it a description "Foobar" (for test 3), and verify that only the three most recent appear in the Report Preview, and a +1 indicator is shown on the last receipt image


This would actually show a +2 indicator because we take into account the other non-receipt requests. Is that expected?

Li357 commented:
> This would actually show a +2 indicator because we take into account the other non-receipt requests. Is that expected?

Great question! The spec says:

> A maximum of three receipt thumbnails will be displayed in the preview component, after which a round counter icon will overlay the right-most thumbnail for each additional request on the report. 

I'm going to clarify, but I believe +2 is the right number. I'll update ASAP!

Li357 commented:
@s77rt Great catch! It should actually be +1. I can make this change.

Li357 commented:
> Bug?: As user B I can see the receipt image in the preview but it's supposed to only be visible to user A

Yep, this is a known backend bug I'll be fixing soon. Essentially, for reports with more than 1 receipt request, whispers will not work until we fix the backend response.

> Bug?: After adding more receipts, user B can see all the receipts and it shows a very confusing message "Only visible to user A" but you as user B can see the receipt as well

This will also be fixed! The reason is `whisperedToAccountIDs` has been removed but Onyx does not reset the property when merging. We'll fix this in the backend response.

> Bug(backend?): The detected amount is negative

I think this a backend bug, yes!

Commit:
Show number of additional scanning receipts in overlay

Commit:
Separate last 3 receipts to display from total number of receipts in an IOU report

Li357 commented:
@s77rt Thank you for your patience here. Turns out there was a logical bug between getting the last three transactions with receipts (which we display) and the total number of transactions with receipts in a report.

Now, we use `ReportUtils.getTransactionsWithReceipts` to search through all the report actions of an IOU report and return those transactions with receipts. This is what `ReportUtils.areAllReceiptsBeingSmartScanned` uses to determine scanning status.

On the other hand, when we display the receipts in the report action item, we use `ReportUtils.getReportPreviewDisplayTransactions` which gets a special property returned by the API that is a list of the last 3 transactions with receipts.

This is not the ideal split in the API but we can fix this later.

s77rt commented on src/components/ReportActionItem/ReportPreview.js:
```diff
@@ -132,46 +162,55 @@ function ReportPreview(props) {
+                        <ReportActionItemImages
+                            images={_.map(lastThreeTransactionsWithReceipts, ({receipt, filename}) => ReceiptUtils.getThumbnailAndImageURIs(receipt.source, filename))}
+                            size={3}
+                            total={ReportUtils.getNumberOfScanningReceipts(props.iouReport)}
```
```suggestion
                            total={transactionsWithReceipts.length}
```

s77rt commented on src/libs/ReportUtils.js:
```diff
@@ -3135,6 +3198,41 @@ function getTaskAssigneeChatOnyxData(accountID, assigneeEmail, assigneeAccountID
+ * @param {Object|null} iouReport
+ * @returns {Number}
+ */
+function getNumberOfScanningReceipts(iouReport) {
+    const reportID = lodashGet(iouReport, 'reportID');
+    const reportActions = ReportActionsUtils.getAllReportActions(reportID);
+    return _.reduce(
+        reportActions,
+        (count, reportAction) => {
+            if (!ReportActionsUtils.isMoneyRequestAction(reportAction)) {
+                return count;
+            }
+            const transaction = TransactionUtils.getTransaction(reportAction.originalMessage.IOUTransactionID);
+            return count + Number(TransactionUtils.hasReceipt(transaction) && TransactionUtils.isReceiptBeingScanned(transaction));
+        },
+        0,
+    );
+}
```
This logic is pretty much the same as `getTransactionsWithReceipts` Let' use that function and filter on `isReceiptBeingScanned`

s77rt commented on src/libs/ReportUtils.js:
```diff
@@ -3135,6 +3198,41 @@ function getTaskAssigneeChatOnyxData(accountID, assigneeEmail, assigneeAccountID
+ */
+function getReportPreviewDisplayTransactions(reportPreviewAction) {
+    const transactionIDs = lodashGet(reportPreviewAction, ['childLastReceiptTransactionIDs'], '').split(',');
+    return _.map(transactionIDs, (transactionID) => TransactionUtils.getTransaction(transactionID));
```
This may cause a crash if the transaction is not there which will return `{}` and we will try to access `receipt.state` on that. Let's use the `hasReceipt` to be safe

Commit:
Remove getNumberOfScanningReceipts

Commit:
Fix header name

Commit:
Run prettier

s77rt commented:
I think attachments view is broken (not related to this PR)

![Screenshot from 2023-08-17 22-20-41](https://github.com/Expensify/App/assets/16493223/ad81a1fa-30f6-4eff-8642-bfb4a9e7febd)

Li357 commented:
@s77rt That's weird, I can't seem to reproduce. We did make some changes to attachments to accomodate receipts but it seems to working fine here. Does console/network show anything?

s77rt commented:
@Li357 I will recheck, it may be related to some local conflicts. I have reproduce the bug on one device (testing device) but not my personal one. Not a blocker either way

s77rt commented:
Bug: I can't upload the receipt on Android

https://github.com/Expensify/App/assets/16493223/e9d67e4d-2684-4afa-98ce-cddc34450963

Commit:
Merge main

s77rt commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/316251c8-defe-45cf-96b7-d30d3f46f9ac


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/54d8152a-76b8-4e84-801e-e994d3d5d6be


</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/caae5b65-5ec1-4eb8-a226-3b6cdf024a0a


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/527ab35e-a559-4195-bac9-e95aad56a98f


</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/38de0041-85e7-491e-9902-605e760eaca8


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

luacmartins commented:
Looks like we're close here. I'll review it again once @s77rt is done with their review

Li357 commented:
I'm having issues with testing, I cannot get Android emulator to make network requests through the app. Asking for another engineer to test this too.

Li357 commented:
Finally able to reproduce this! Looking into it!

Commit:
Fix weird Android image upload issue

Li357 commented on src/libs/ReportUtils.js:
```diff
@@ -1245,6 +1245,50 @@ function getTransactionDetails(transaction) {
+ * @param {Object|null} iouReport
+ * @returns {[Object]}
+ */
+function getTransactionsWithReceipts(iouReport) {
```
WiIl fix in follow up!

Li357 commented:
@s77rt I think I fixed the issue? Seems like due to an Android bug, `res` actually contains the image, just that `res.ok` and `res.statusText` are not set on Android leading to the throw. I added a check if on Android as a workaround.

Li357 committed:
Merge branch 'main' into andrewli-displayreceipts

s77rt commented:
Maybe a backend error? The `childLastReceiptTransactionIDs` have 2 transactions only but there are 3.

![Screenshot from 2023-08-19 03-33-11](https://github.com/Expensify/App/assets/16493223/99f2adb3-55e2-410f-b370-39158034c444)

s77rt commented:
The preview image in Android does not load

https://github.com/Expensify/App/assets/16493223/b93c8b1a-f460-4c75-a721-2222cb2cb425

Commit:
Use optimistic transaction in updating report preview

Li357 commented:
@s77rt Are you experiencing the same API problems on Android? Seems like it's not loading in the emulator. Regardless, I've updated the optimistic report preview to show the right receipt image (regression from merging main).

s77rt commented:
The preview image is not updated instantly, only after changing report - May be related to https://github.com/Expensify/App/pull/24235#issuecomment-1684699445


https://github.com/Expensify/App/assets/16493223/1567dbb8-776d-449a-8138-111340f8c9ad


https://github.com/Expensify/App/assets/16493223/aa9a2a53-58b2-4b67-9402-1b8eb01606da

Li357 commented:
I suspect this is a backend Onyx issue where we need to send an additional update. I'm going to address a few backend changes in a PR on Monday 👍

s77rt commented:
@Li357 Should hold testing on Android till this is resolved?

Li357 commented:
I don't think so. I feel the more we wait on this harder it's going to be to merge cleanly given the work that's being done in parallel.

s77rt commented:
@Li357 Can you complete the author checklist? - I have added a video for Android.

Li357 committed:
Merge branch 'main' into andrewli-displayreceipts

Li357 commented:
@s77rt I've updated the checklist. I'm unable to attach screenshots for Android because my emulator is still giving me trouble and neither local nor staging work for me behind a VPN. Some other internal engineers are facing this issue, but I'm going to get another to test things here!

Li357 commented:
For tracking, here are a list of issues to address in a follow-up:

- Refactor ReportActionItemImage to not use `RenderHTML` and remove w100, h100 global styling changes, remove data-expensify-fit-container
- Pass ID to getTransactionsWithReceipts
- Fix backend issue with childLastReceiptTransactionIDs

s77rt commented:
> Verify that clicking on the receipt image in this view opens up a modal with title "Receipt"

If offline we can't view the image. Is that expected?

https://github.com/Expensify/App/assets/16493223/d9cf738d-9290-4b41-8a02-f05ff7e15c3a

Li357 commented:
@s77rt I will double check, but either way I would say that's out of scope for this PR.

s77rt reviewed (approved):
LGTM! :rocket:

s77rt commented:
@Li357 Let me know your findings regarding https://github.com/Expensify/App/pull/24235#issuecomment-1682997204 ([slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1692630012291809?thread_ts=1692627732.848999&cid=C01GTK53T8Q))

Li357 commented:
Unassigning you @joelbettner since I think we've got this covered for internal reviewers 😄

mountiny commented on src/components/Attachments/AttachmentCarousel/extractAttachmentsFromReport.js:
```diff
@@ -36,6 +39,23 @@ function extractAttachmentsFromReport(report, reportActions) {
+        // We're handling receipts differently here because receipt images are not
+        // part of the report action message, the images are constructed client-side
+        if (ReportActionsUtils.isMoneyRequestAction(action)) {
+            const transaction = TransactionUtils.getTransaction(action.originalMessage.IOUTransactionID);
```
I think that not every IOU action has the `originalMessage.IOUTransactionID`, could we access it safely and return early if empty?

luacmartins replied:
+1 pay actions don't have the `transactionID`. Let's fix it in this PR to avoid any crashes.

Commit:
Access transactionID safely in case of pay IOU actions

Commit:
Fix lint

Li357 committed:
Merge branch 'main' into andrewli-displayreceipts

Commit:
Fix lint

luacmartins commented:
Web PR is in prod. We got 3 approvals. Let's :shipit:

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

Li357 commented on src/libs/ReportUtils.js:
```diff
@@ -1248,6 +1248,50 @@ function getTransactionDetails(transaction) {
+function areAllRequestsBeingSmartScanned(iouReport, reportPreviewAction) {
+    const transactions = getTransactionsWithReceipts(iouReport);
+    // If we have more requests than requests with receipts, we have some manual requests
+    if (ReportActionsUtils.getNumberOfMoneyRequests(reportPreviewAction) > transactions.length) {
```
Yep! The one thing we need to do is update childMoneyRequestAmount on the backend.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5931749076) to staging by https://github.com/luacmartins in version: 1.3.56-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

mvtglobally commented:
@Li357 @luacmartins @mountiny Do we need to QA this one or no? 
![Screen Shot 2023-08-21 at 10 55 05 PM](https://github.com/Expensify/App/assets/43995119/de01852e-7a57-4153-a771-d421e3522cf3)

situchan commented:
It seems many regressions came from this PR:
- https://github.com/Expensify/App/issues/25616
- https://github.com/Expensify/App/issues/25640
- https://github.com/Expensify/App/issues/25644
- https://github.com/Expensify/App/issues/25648

Better to revert?

Li357 commented:
@mvtglobally I'll update this! QA is same as the listed steps under Tests

Li357 commented:
@situchan This PR unblocks a lot of internal issues so I don't think a revert is good here. Let's fix these linked issues ASAP!

shawnborton commented:
Hey team - just a friendly reminder, when we're merging lots of UI changes like this, please please please tag the design team for review. I already see a few things we missed that will need to be fixed in a follow up PR. 

For example, on the scan banner, why is the text on the right right-aligned? 

Or on this screen, why are we using a full-width thumbnail?
<img width="820" alt="image" src="https://github.com/Expensify/App/assets/2319350/8de1d9ff-a2c1-4f0b-976e-679e18540712">

cc @JmillsExpensify @trjExpensify

shawnborton commented:
Yup, mind tagging me in the PR and I'll give it a look?

dangrous commented:
Hi! According to `git bisect`, this caused a Deploy Blocker [here](https://github.com/Expensify/App/issues/25678#issuecomment-1688235919) - since this is a big PR, I could use a hand figuring out what happened and what to fix, it looks like it's this commit:

```
commit 941d730532ec090e4e4e2f5d91d0388f1ac63864
Author: Andrew Li <andrewli@expensify.com>
Date:   Wed Aug 9 09:29:09 2023 -0400

    Fit images to container

 src/components/HTMLEngineProvider/BaseHTMLEngineProvider.js   |  2 +-
 .../HTMLEngineProvider/HTMLRenderers/ImageRenderer.js         | 11 ++++++++---
 src/components/ReportActionItem/ReportPreview.js              |  7 ++++---
 src/components/ThumbnailImage.js                              |  8 +++++++-
 src/styles/styles.js                                          |  2 +-
 5 files changed, 21 insertions(+), 9 deletions(-)
```

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5958899174) to production by https://github.com/roryabraham in version: 1.3.56-24 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

Li357 commented on src/components/ReportActionItem/MoneyRequestPreview.js:
```diff
@@ -189,64 +205,78 @@ function IOUPreview(props) {
-                                    isHovered={props.isHovered}
-                                    shouldUseCardBackground
-                                />
+                        {requestMerchant && (
```
Caused this regression: https://github.com/Expensify/App/issues/25616

luacmartins commented:
This PR introduced [this bug](https://github.com/Expensify/App/issues/26590) because we didn't wrap images without thumbnail in a pressable component

cubuspl42 commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -469,10 +469,21 @@ function ReportActionItem(props) {
+                        checkIfContextMenuActive: toggleContextMenuFromActiveReportAction,
+                    }}
+                >
+                    <OfflineWithFeedback pendingAction={props.action.pendingAction}>
```
Coming from https://github.com/Expensify/App/issues/26235

This line brought back wrapping `MoneyRequestView` in `OfflineWithFeedback`, which was removed in 
https://github.com/Expensify/App/pull/25399. This might be a mistake in merge conflict resolution.

This caused a double application of the opacity reduction effect.

cubuspl42 commented:
Coming from https://github.com/Expensify/App/issues/27697

It was missed that the "Receipt scan in progress…" message doesn't always display in the report preview

rushatgabhane commented on src/components/ReportActionItem/MoneyRequestPreview.js:
```diff
@@ -177,6 +185,14 @@ function IOUPreview(props) {
         return message;
     };
 
+    const getDisplayAmountText = () => {
```
Hi, we didn't handle offline behavior for this function. It defaulted to showing amount as $0.

https://github.com/Expensify/App/issues/29048#issuecomment-1751681314

eh2077 commented on src/languages/en.js:
```diff
@@ -399,6 +401,11 @@ export default {
         viewDetails: 'View details',
         pending: 'Pending',
+        deleteReceipt: 'Delete receipt',
+        receiptScanning: 'Receipt scan in progress…',
```
We should handle plural use case when there're multiple receipts. It caused https://github.com/Expensify/App/issues/49054
